vyos/vyos-1x
# sshguard: T5354: Add service ssh dynamic-protection

## Change Summary

Sshguard protects hosts from brute-force attacks.
It can inspect logs and block "bad" addresses by threshold Auto-generates own tables and rules for nftables, so they are not intercepted with VyOS firewall rules.
When service stops, all generated tables are deleted.

## Types of changes


## Related Task(s)

* https://vyos.dev/T5354

## Component(s) name

ssh, sshguard
## Proposed changes


## How to test
```
set service ssh dynamic-protection allow-from '192.0.2.1'
set service ssh dynamic-protection block-time '120'
set service ssh dynamic-protection detect-time '1800'
set service ssh dynamic-protection threshold '30'

```
Try several attempts to login with incorrect login/pass and check nftables
```
table ip sshguard {
	set attackers {
		type ipv4_addr
		flags interval
		elements = { 192.168.122.14 }
	}

	chain blacklist {
		type filter hook input priority filter - 10; policy accept;
		ip saddr @attackers drop
	}
}
table ip6 sshguard {
	set attackers {
		type ipv6_addr
		flags interval
	}

	chain blacklist {
		type filter hook input priority filter - 10; policy accept;
		ip6 saddr @attackers drop
	}
}

```
Smoketest:
```
vyos@r11:~$ /usr/libexec/vyos/tests/smoke/cli/test_service_ssh.py
test_ssh_default (__main__.TestServiceSSH) ... ok
test_ssh_dynamic_protection (__main__.TestServiceSSH) ... ok
test_ssh_login (__main__.TestServiceSSH) ... /usr/lib/python3/dist-packages/paramiko/kex_ecdh_nist.py:39: CryptographyDeprecationWarning: encode_point has been deprecated on EllipticCurvePublicNumbers and will be removed in a future version. Please use EllipticCurvePublicKey.public_bytes to obtain both compressed and uncompressed point encoding.
  m.add_string(self.Q_C.public_numbers().encode_point())
/usr/lib/python3/dist-packages/paramiko/kex_ecdh_nist.py:96: CryptographyDeprecationWarning: Support for unsafe construction of public numbers from encoded data will be removed in a future version. Please use EllipticCurvePublicKey.from_encoded_point
  self.curve, Q_S_bytes
/usr/lib/python3/dist-packages/paramiko/kex_ecdh_nist.py:111: CryptographyDeprecationWarning: encode_point has been deprecated on EllipticCurvePublicNumbers and will be removed in a future version. Please use EllipticCurvePublicKey.public_bytes to obtain both compressed and uncompressed point encoding.
  hm.add_string(self.Q_C.public_numbers().encode_point())
ok
test_ssh_multiple_listen_addresses (__main__.TestServiceSSH) ... ok
test_ssh_single_listen_address (__main__.TestServiceSSH) ... ok
test_ssh_vrf (__main__.TestServiceSSH) ... ok

----------------------------------------------------------------------
Ran 6 tests in 33.605s

OK

```

## Checklist:
